File: LintCode/java/BinaryTreePostorderTraversal.java
import java.util.ArrayList;

/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class BinaryTreePostorderTraversal {
    /**
     * @param root: The root of binary tree.
     * @return: Preorder in ArrayList which contains node values.
     */

    // V1, O(n)
    // Recursion
    public ArrayList<Integer> preorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<Integer>();
        traverse(root, result);
        return result;
    }

    private void traverse(TreeNode root, ArrayList<Integer> result) {
        if (root == null) {
            return;
        }

        traverse(root.left, result);
        traverse(root.right, result);
        result.add(root.val);
    }

    // V2, O(n)
    // Recursion
    public ArrayList<Integer> postorderTraversal(TreeNode root) {
        ArrayList<Integer> result = new ArrayList<Integer>();
        if (root == null) {
            return result;
        }

        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode prev = null;
        TreeNode curr = null;

        while (root != null || !stack.isEmpty()) {
            if (root != null) {
                stack.push(root);
                root = root.left;
            } else {
                curr = stack.peek();
                if (curr.right != null && curr.right != prev) {
                    root = curr.right;
                } else {
                    stack.pop();
                    result.add(curr.val);
                    prev = curr;
                }
            }
        }

        return result;
    }
}

File: LeetCode-Python/binary_tree_postorder_traversal.py
# Given a binary tree, return the postorder traversal of its nodes' values.

# For example:
# Given binary tree {1,#,2,3},
#    1
#     \
#      2
#     /
#    3
# return [3,2,1].

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def postorderTraversal_1_recursive(self, root):
        res = []
        self.postorder_traversal_recursive(root, res)
        return res

    def postorder_traversal_recursive(self, root, res):
        if root == None: return
        self.postorder_traversal_recursive(root.left, res)
        self.postorder_traversal_recursive(root.right, res)
        res.append(root.val)
        return res


    def postorderTraversal_2_iterative(self, root):
        if root == None: return []
        res = []
        stack = []
        last_node_visited = None
        while root or stack:
            if root:
                stack.append(root)
                root = root.left
            else:
                peek_node = stack[-1]
                if peek_node.right and last_node_visited != peek_node.right:
                    root = peek_node.right
                else:
                    stack.pop()
                    res.append(peek_node.val)
                    last_node_visited = peek_node
        return res